runtime/ioext/asyncpipe.go

```diff
@@ -0,0 +1,128 @@
+package ioext
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+// ErrPipeFull is returned from AsyncPipeWriter.Write if the pipes capacity has
+// been reached.
+var ErrPipeFull = errors.New("The internal pipe buffer have reached its capacity")
+
+// AsyncPipeReader is a reading side of an AsyncPipe, similar to io.PipeReader.
+type AsyncPipeReader struct {
+	*AsyncPipeWriter
+}
+
+// AsyncPipeWriter is a writing side of a BlockedPipe, similar to io.PipeWriter.
+type AsyncPipeWriter struct {
+	c        sync.Cond
+	m        sync.Mutex
+	buffer   []byte
+	closed   bool
+	broken   bool
+	tell     chan<- int
+	capacity int
+}
+
+// AsyncPipe is similar to io.Pipe() except that writes isn't blocking, instead
+// data will be added to an internal buffer that can grow up to specified
+// capacity.
+// Additionally, you may supply a channel tell that will be told whenever
+// N bytes have been read, so that more bytes can be requested to be written.
+//
+// This pipe kind is useful when implementing congestion control.
```

Comment:
I think your `AsyncPipeWriter` could be achieved with a regular `byte` channel with the given `capacity`. This would certainly be more idiomatic, less code to manage/maintain, easier for contributors to reason about (as it would use standard low-level go mechanics) and probably would be similarly efficient. What are the drawbacks of using a regular channel with the given capacity?